bench/irmin-pack/tree.ml

```diff
@@ -90,247 +218,528 @@ module Parse_trace = struct
+      { path; channel; count = 0L }
 
-  let populate_array ncommits flatten path =
-    let commits = Array.init ncommits (fun _ -> []) in
-    let n = read_commits commits ncommits flatten path in
-    (commits, n)
+    let create cache_dir =
+      let tbl =
+        op_tags
+        |> List.map (fun op -> (op, create_op cache_dir op))
+        |> List.to_seq
+        |> Hashtbl.of_seq
+      in
+      Hashtbl.find tbl
+
+    let cleanup t = List.iter (fun op -> Sys.remove (t op).path) op_tags
+
+    let write_duration : out_channel -> float -> unit =
+      let encode_duration = Repr.(encode_bin int32 |> unstage) in
+      fun channel d ->
+        encode_duration (Int32.bits_of_float d) (output_string channel)
+
+    let push_duration t op d : unit =
+      let t = t op in
+      t.count <- Int64.add t.count 1L;
+      write_duration t.channel d
+
+    let read_duration : in_channel -> float =
+      let decode_duration = Repr.(decode_bin int32 |> unstage) in
+      fun channel ->
+        let s = really_input_string channel 4 in
+        decode_duration s 0 |> snd |> Int32.float_of_bits
+
+    let open_duration_sequence path count : float Seq.t =
+      let channel = open_in path in
+      assert (LargeFile.in_channel_length channel = Int64.mul count 4L);
+      let aux i =
+        if i >= count then None else Some (read_duration channel, Int64.add i 1L)
+      in
+      Seq.unfold aux 0L
+
+    (** The computed summaries contain 3 informations for each operations.
+
+        {3 Histograms}
+
+        The [histo] field is computed using https://github.com/barko/bentov.
+
+        [Bentov] computes dynamic histograms without the need for a priori
+        information on the distributions, while maintaining a constant memory
+        space and a marginal CPU footprint.
+
+        The implementation of that library is pretty straightforward, but not
+        perfect; it doesn't scale well with the number of bins.
+
+        The computed histogram depends on the order of the operations, some
+        marginal unsabilities are to be expected.
+
+        [Bentov] is good at spreading the bins on the input space. Since these
+        data will be shown on a log plot, the log10 of those values is passed to
+        [Bentov] instead, but the json will store real seconds.
+
+        {3 Moving Averages}
+
+        [ma_xs] and [ma_ys] form a smoothed curve describing the variations over
+        time of the call durations to an op.
+
+        [ma_xs] is an increasing list of float. The last element is [1.0] and
+        the first is [1 / moving_average_sample_count].
+
+        [ma_ys] is a list of call durations smoothed using an exponential decay
+        moving average, defined from its half life through the
+        [moving_average_half_life_ratio] constant.
+
+        {3 Max Values}
+
+        The [max_point] field contains the value and the index of the longest
+        occurence of an operation. *)
+    module Summary = struct
```

Comment:
what do you think of moving this in a separate file? I did not follow everything you did here, but maybe it is general enough that it can be decoupled from the rest of the file ?

Reply:
I didn't split the file yet because I'm unsure if only `Bootstrap_trace.Stats` should be moved out, or `Bootstrap_trace` + `Generate_trees_from_trace`.

Reply:
I would move only `Bootstrap_trace.Stats`, but either way is fine.